### ISSUE TITLE
remove effect question

### ISSUE THREAD

**Issue by zjun0214** [beat-machine/beat-webapp#38]
(write by using translate machine)

Thank you for making wonderful program.
I’d like to ask you a few questions.

I understand remove effect

-
Input :
  1 2 3 4 5 6 7 8 ...

After remove every '2' :
  1 2 / 5 6 / 7 8 / 11 12 / 13 14 ...
-

like this, but

-
Input :
  1 2 3 4 5 6 7 8 ...

After remove every ?????? :
  3 4 / 7 8 / 11 12 / 15 16 / 19 20 ...
-

Thank you for making wonderful program.
I’d like to ask you a few questions.

I understand remove effect

-
Input :
  1 2 3 4 5 6 7 8 ...

After remove every 2 :
  1 2 / 5 6 / 7 8 / 11 12 / 13 14 ...
-

like this, but

How can I make song beat

-
Input :
  1 2 3 4 5 6 7 8 ...

After remove every ???😳 :
  3 4 / 7 8 / 11 12 / 15 16 / 19 20 ...
-

How can I make song beats like this??

**Comment by branchpanic** [comment]
Thanks for your interest! When you add multiple effects, they get run one after another. I was able to approximate your example using the following.

1. Remove every 4 with offset 3
2. Remove every 3 with offset 3

Here's what each step of the process looks like.

1. Input
    [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
2. After removing every 4 with offset 3
    [1, 2, 3, 4, 7, 8, 10, 12, 14, 15, 18, 19, 20]
3. After removing every 3 with offset 3
    [1, 2, 3, 4, 7, 8, 11, 12, 15, 16, 19, 20]

There's one caveat, which is that the first few beats are unaffected. This was originally intended behavior, but I now realize it's a bit confusing. I plan on fixing it in https://github.com/beat-machine/beat-machine/issues/44.

**Comment by JustACuteGirlLookingForAManToFixHerBugs** [comment]
> * Input
>   [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
> * After removing every 4 with offset 3
>   [1, 2, 3, 4, 7, 8, 10, 12, 14, 15, 18, 19, 20]
> * After removing every 3 with offset 3
>   [1, 2, 3, 4, 7, 8, 11, 12, 15, 16, 19, 20]

It would be soo helpful if the program would actually display numbers like you're showing here when adding effects. Plus, this way it would be possible to debug issues with more certainty.

**Comment by branchpanic** [comment]
Totally agree, I wish I had more time to implement even a prototype version of this in #17. It's probably the biggest problem right now in terms of complexity.

**Comment by JustACuteGirlLookingForAManToFixHerBugs** [comment]
Is it more complicated than it seems at first sight? I don't have much development experience, so it's hard for me to get the right perspective, but for me it seems like the beat machine itself is like several orders of magnitude more complex than the visualization.

**Comment by JustACuteGirlLookingForAManToFixHerBugs** [comment]
The Cut effect is probably more challenging to present visually, but it could be totally omitted in the first version. I bet the swap and remove are two most popular effects used, so covering them would go a long way.

**Comment by branchpanic** [comment]
You're right that it really wouldn't be that bad, my main problems are writing a decent frontend and actually getting around to it amidst work and other projects.

**Comment by JustACuteGirlLookingForAManToFixHerBugs** [comment]
Btw, what else are you working on? Is it something fun as well?

**Comment by branchpanic** [comment]
Mostly graphics stuff in Unity since it's directly related to my field, but I'm occasionally doing other media projects like [a video corruptor](https://github.com/branchpanic/vdcrpt).

I'm hoping to get all the beat machine stuff into a state where someone else can readily pick it up. That way I can focus on the former. Getting off-topic now, I should make a post somewhere to clarify I'm not working on this project super actively.
